demo/popcha-client/web/src/main.py

```diff
@@ -0,0 +1,222 @@
+    :return: True if the configurations is valid
+    """
+    if "client_id" not in config.keys() or config["client_id"] == '':
+        config["client_id"] = secrets.token_urlsafe(32)
+        config_file.seek(0)
+        config_file.write(json.dumps(config))
+        return False
+    if "host_url" not in config.keys() or config["host_url"] == "":
+        raise ValueError(
+                "The \"server_url\" property should be set in "
+                "config.json and not empty"
+                )
+    if "host_port" not in config.keys() or config["host_port"] < 1:
+        raise ValueError(
+                "The \"server_port\" should be set in config.json "
+                "and greater than 0"
+                )
+    return True
+
+
+def check_provider(provider: dict) -> bool:
+    """
+    Check that a single provider is valid. It only performs static checks and
+    no connectivity tests.
+    :param provider: The provider to check
+    :return: True if the provider is valid
+    """
+    valid_lao_id = (("lao_id" in provider) and (len(provider["lao_id"]) == 44)
+                    and provider["lao_id"].endswith("="))
+    valid_domain = (("domain" in provider) and ('/' not in provider["domain"])
+                    and not provider["domain"].startswith("http"))
+    valid_public_key = len(provider["public_key"]) > 0
+    return valid_lao_id and valid_domain and valid_public_key
+
+
+def filter_providers(providers: list) -> list:
+    """
+    Keeps only the providers that are valid from the original list
+    :param providers: The original list of providers
+    :return: The valid providers of the original list
+    """
+    return [provider for provider in providers if check_provider(provider)]
+
+
+def load_providers() -> list:
+    """
+    Loads the providers from the configuration file and returns them
+    :return: The providers present in the configuration files
+    """
+    provider_path = path.normpath(
+        path.join(
+            path.dirname(__file__),
+            "../data/providers.json"
+            )
+        )
+    with open(provider_path) as provider_file:
+        return json.loads(provider_file.read())
+
+
+def on_startup() -> None:
+    """
+    Prepare the data needed by the client authentication server (This
+    server). This includes the list of authorized
+    Open ID providers.html, as well as basic config information such as the
+    homepage HTML code or the
+    """
+    global config, core_app, authenticationProvider
+    csrf = CSRFProtect()
+    csrf.init_app(app)
+    core_app = CounterApp()
+    authenticationProvider = Authentication()
+    providers = load_providers()
+    authenticationProvider.providers = filter_providers(providers)
+    config_path = path.normpath(
+        path.join(
+            path.dirname(__file__),
+            "../data/config.json"
+            )
+        )
+    with open(config_path, "r+") as config_file:
```

Comment:
What if the file doesn't exit?

Reply:
Then the process will imedidatly crash with an `FileNotFoundError: [Errno 2] No such file or directory:` error. According to this : [https://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c](https://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c), there is no clean easy way to stop the server on failure, so I think that the current behavior is enough.

Reply:
I thought it might be nice to show an error but I guess `FileNotFoundError` is explicit enough :)